File: leetcode/go/next-permutation.go
package main

import "fmt"

/*
https://leetcode.com/problems/next-permutation
 */

func nextPermutation(nums []int) {
	n := len(nums)
	i := n - 1
	for ; i > 0; i-- {
		if nums[i] > nums[i-1] {
			break
		}
	}
	if i != 0 {
		j := i
		for ; j<n-1; j++ {
			if nums[j+1] <= nums[i-1] {
				break
			}
		}
		nums[i-1], nums[j] = nums[j], nums[i-1]
	}
	for j, k := i, n-1; j < k; j++ {
		nums[j], nums[k] = nums[k], nums[j]
		k--
	}
}

func main() {
	for _, nums := range [][]int{{1,2,0,3,7,6,5}, {2,3,1}, {1,5,1}} {
		nextPermutation(nums)
		fmt.Println(nums)
	}
}

File: leetcode/next_permutation.py
class Solution(object):
    def nextPermutation(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        sz = len(nums)
        if sz == 0 or sz == 1:
            return

        for i in range(sz-1, -1, -1):
            if i > 0 and nums[i-1] < nums[i]:
                break

        nums[i:] = nums[i:][::-1]

        for j in range(i, sz):
            if nums[i-1] < nums[j]:
                break
        nums[i-1], nums[j] = nums[j], nums[i-1]
        return

if __name__ == '__main__':
    cases = [
        [2,1,5,3,4],
        [2,1,3,4,5,6],
        [2,1,6,5,4,2],
        [2,5,6,5,4,2],
        [2,5,6,4,3,2],
        [3,2,1]
    ]

    solution = Solution()
    for case in cases:
        print(case)
        (solution.nextPermutation(case))
        print(case)

